src/main/java/picard/illumina/BarcodeExtractor.java

```diff
@@ -0,0 +1,248 @@
+import java.util.concurrent.ConcurrentHashMap;
+
+public class BarcodeExtractor {
+    final Map<String, BarcodeMetric> metrics;
+    final Set<ByteString> barcodesBytes;
+    final BarcodeMetric noMatch;
+    final int maxNoCalls, maxMismatches, minMismatchDelta, minimumBaseQuality;
+    final DistanceMetric distanceMode;
+    final ConcurrentHashMap<ByteString, BarcodeMatch> barcodeLookupMap;
+
+    public BarcodeExtractor(final Map<String, BarcodeMetric> barcodeToMetrics,
+                            final ConcurrentHashMap<ByteString, BarcodeMatch> barcodeLookupMap,
+                            Set<ByteString> barcodeByteStrings, final BarcodeMetric noMatchMetric,
+                            final int maxNoCalls,
+                            final int maxMismatches,
+                            final int minMismatchDelta,
+                            final int minimumBaseQuality,
+                            final DistanceMetric distanceMode) {
+        this.metrics = barcodeToMetrics;
+        this.barcodeLookupMap = barcodeLookupMap;
+        this.barcodesBytes = barcodeByteStrings;
+        this.noMatch = noMatchMetric;
+        this.maxNoCalls = maxNoCalls;
+        this.maxMismatches = maxMismatches;
+        this.minMismatchDelta = minMismatchDelta;
+        this.minimumBaseQuality = minimumBaseQuality;
+        this.distanceMode = distanceMode;
+    }
+    /**
+     * Find the best barcode match for the given read sequence, and accumulate metrics
+     *
+     * NOTE: the returned BarcodeMatch object will contain mismatches mismatchesToSecondBest values that may be
+     * inaccurate as long as the conclusion match/no-match isn't affected. for example, mismatches and mismatchesToSecondBest
+     * may be smaller than their true value if mismatches is truly larger than maxMismatches.
+     * Also, mismatchesToSecondBest might be smaller than its true value if its true value is greater than
+     * mismatches + minMismatchDelta. This is due to an optimization which allows the distance calculation to stop once
+     * the conclusion (Match or no-Match) can be reached.
+     *
+     * @param readSubsequences portion of read containing barcode
+     * @return perfect barcode string, if there was a match within tolerance, or null if not.
+     */
+    BarcodeMatch findBestBarcode(final byte[][] readSubsequences,
+                                 final byte[][] qualityScores) {
+        final boolean canUseLookupTable = areAllQualitiesAboveMinimum(qualityScores, minimumBaseQuality);
+        final ByteString barcodesAsString = new ByteString(readSubsequences);
+        BarcodeMatch match = canUseLookupTable ? barcodeLookupMap.get(barcodesAsString) : null;
+        if (match == null) {
+            match = calculateBarcodeMatch(readSubsequences, qualityScores);
+
+            if (canUseLookupTable && match.isMatched()) {
+                barcodeLookupMap.put(barcodesAsString, match);
+            }
+        }
+
+        return match;
+    }
```

Comment:
```suggestion
        final boolean canUseLookupTable = areAllQualitiesAboveMinimum(qualityScores, minimumBaseQuality);
        if (canUseLookupTable) {
            final ByteString barcodesAsString = new ByteString(readSubsequences);
            BarcodeMatch match = barcodeLookupMap.get(barcodesAsString);
            if (match == null) match = calculateBarcodeMatch(readSubsequences, qualityScores);
            if (match.isMatched()) barcodeLookupMap.put(barcodesAsString, match);
            return match;
        }
        else {
            return calculateBarcodeMatch(readSubsequences, qualityScores);
        }
    }
```

To avoid making the ByteString if we can't use it, and tidy up a bit.